Browse Diff tab menu alternatives and presentation

gerhardol opened the pull request:
Fixes #4564
Fixes #4387
Fixes #4396

Note: Due to the structure of tests in a separate class, it is very difficult to split this issue in several commits. First commit was PR #4566, partly reviewed. (Depended on merged commits #4562 #4563 #4565 #4567 #4568)


Tests:
Rewrote RevisionDiffController to make it possible to have some kind of tests. The tests are retrofitted to the current functionality as they are just testing the menu items and not the actions themselves. (The tests adds may more maintenance than they give benefits but the formal test coverage increases.)

Browse Diff:
Submodules actions available for multi select (#4568)
Presents the Reset options for the correct parent in multi select situations
No longer differ between parent-child and first-second for Reset
Difftool: Describes A and B revisions in the menu
Difftool arguments depends on parent availability
Better detection of parents to A/B for diffs

Commit:
Limit DiffTool and FileHistory to tracked

FormDiff:
Limit FileHistory to Tracked
Use RevisionDiffController to align to Browse-Diff for difftool
 
Screenshots before and after (if PR changes UI):
-
![image](https://user-images.githubusercontent.com/6248932/36948593-60e9560a-1fdd-11e8-9206-d77d8acb150b.png)
![image](https://user-images.githubusercontent.com/6248932/36948595-7524f2be-1fdd-11e8-918e-078c00545647.png)
![image](https://user-images.githubusercontent.com/6248932/36948620-abd497ce-1fdd-11e8-8549-d4246dd619a4.png)


![image](https://user-images.githubusercontent.com/6248932/36948634-019fa432-1fde-11e8-92a9-b8dbb5fcc422.png)
![image](https://user-images.githubusercontent.com/6248932/36948644-2d5d6b54-1fde-11e8-8da3-0e3e2f9379b9.png)
![image](https://user-images.githubusercontent.com/6248932/36948629-df36db72-1fdd-11e8-8941-a78997bf361e.png)


What did I do to test the code and ensure quality:
 - Added tests
 - Manual testing

Has been tested on (remove any that don't apply):
 - Windows 10

gerhardol commented:
The first two commits are being reviewed in #4565, the third was mostly reviewed in #4566 (but cannot be merged in isolation). The review is primarily for the last commit - can be done after #4565 is merged.

RussKie commented on GitUI/CommandsDialogs/FormDiff.cs:
```diff
@@ -240,14 +249,47 @@ private void btnAnotherHeadCommit_Click(object sender, EventArgs e)
+        private ContextMenuDiffToolInfo GetContextMenuDiffToolInfo()
+        {
+            bool firstIsParent = _revisionDiffController.IsFirstParent(DiffFiles.Revision.ParentGuids, DiffFiles.SelectedItemParents.Select(i => i.Guid));
+            bool localExists = _revisionDiffController.LocalExists(DiffFiles.SelectedItemsWithParent, _fullPathResolver);
```
this indicates an incorrect use of `IFullPathResolver` - behaviors must be injected via constructors and not passed as parameters

gerhardol replied:
I see two two solutions for the relative simple LocalExists() function:
 * Add LocalExists() to IFullPathResolver instead
 *  Duplicate LocalExists() where it is used (two or three modules). A separate interface/class for this seem to be overkill.

RussKie replied:
> Add LocalExists() to IFullPathResolver instead

:+1:

gerhardol replied:
I duplicated LocalExists() as it deals with GitItemStatusWithParent.
Probably better to move other similar to a separate class (but FullPathResolver.Resolve() will be duplicated instead).

RussKie commented:
I've pulled your branch down, I'll send a tweak shortly

On 13/03/2018 8:04 AM, "Gerhard Olsson" <notifications@github.com> wrote:

> *@gerhardol* commented on this pull request.
> ------------------------------
>
> In GitUI/CommandsDialogs/FormDiff.cs
> <https://github.com/gitextensions/gitextensions/pull/4569#discussion_r173946135>
> :
>
> >              bool isExactlyOneItemSelected = DiffFiles.SelectedItems.Count() == 1;
> -            blameToolStripMenuItem.Visible = isExactlyOneItemSelected && !(DiffFiles.SelectedItem.IsSubmodule || _baseRevision.IsArtificial);
> +
> +            openWithDifftoolToolStripMenuItem.Enabled = isAnyTracked;
> +            fileHistoryDiffToolstripMenuItem.Enabled = isAnyTracked && isExactlyOneItemSelected;
> +            blameToolStripMenuItem.Enabled = fileHistoryDiffToolstripMenuItem.Enabled && !DiffFiles.SelectedItem.IsSubmodule;
> +        }
> +
> +        private ContextMenuDiffToolInfo GetContextMenuDiffToolInfo()
> +        {
> +            bool firstIsParent = _revisionDiffController.IsFirstParent(DiffFiles.Revision.ParentGuids, DiffFiles.SelectedItemParents.Select(i => i.Guid));
> +            bool localExists = _revisionDiffController.LocalExists(DiffFiles.SelectedItemsWithParent, _fullPathResolver);
>
> I duplicated LocalExists() as it deals with GitItemStatusWithParent.
> Probably better to move other similar to a separate class (but
> FullPathResolver.Resolve() will be duplicated instead).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/4569#discussion_r173946135>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXo6nMos4TjM3SRMZQTeQzDUDHeD3ks5tduL2gaJpZM4SbbcL>
> .
>

RussKie commented:
Sorry, my laptop died on a way to work so I will sent a tweak later tonight

RussKie commented:
I've have taken liberty and shuffled things around.

Referencing `RevisionDiffController` in places other than `RevisionDiff` is incorrect (I must have missed it when it happened). 

I've extracted the common functionality:
* what I understood was related to testing/checking of `GitRevision` - I moved into `GitRevisionTester`. I'm not 100% settled, but it feels more appropriate than having that logic in `RevisionDiffController` and then referenced it somewhere else.
* shared context menu logic I moved out to `RevisionDiffContextMenuController`. I couldn't come up with a better name, so feel free to suggest alternatives.

Thought? Opinions?

gerhardol commented:
I am OK with the changes. Will suggest a few name changes, will readd description etc
Testcases are still quite meaningless, but it looks good with increased coverage.

LocalRevisionExists -> AnyLocalFileExists()

IsFirstParent()
It makes more sense to have FileStatusList as an argument here, that is what is being checked.
But then GitRevisionTester is no longer a fit for the method.

> Referencing RevisionDiffController in places other than RevisionDiff is incorrect (I must have missed it when it happened).

That was done here

> shared context menu logic I moved out to RevisionDiffContextMenuController.

FileStatusListController or FileStatusListContextMenuController would be more logical. 
Remaining parts of RevisionDiffController fits there too even if currently only used in RevisionDiff.cs.

RussKie commented:
> Testcases are still quite meaningless

Disagree. We see a lot of changes coming at a high rate (no one is working here full time). 
Tests help to 
* document design decisions - a good test name spells out requirements (e.g. `MyMethod_should_do_something_in_case_of_bla`), and 
* guard against possible regressions.

Speaking of which, to me the logic in both new methods looked very odd. 


> > shared context menu logic I moved out to RevisionDiffContextMenuController.
> 
> FileStatusListController or FileStatusListContextMenuController would be more logical.

No probs, let's rename. I like `FileStatusListContextMenuController` sightly more.

> Remaining parts of RevisionDiffController fits there too even if currently only used in RevisionDiff.cs.

The reason I created the controller (and other <UI control>Controllers) is to abstract business logic from the UI layer. I have a dream that one day we may be able to use another UI framework, so the UI should do as little as possible.
So I need to think more about this proposal, but generally unless/until it is used in multiple I see no reason to share it.

gerhardol commented:
Renamed and updated. Did not move more around.
Will need to test a little more myself.

> Disagree. 

We have had this discussion before and try my best to comply to your vision. But in this situation the tests just takes time and decrease my motivation. These tests are just at the "looks" not at the actual contents. The functionality has to broken down considerably different to meaningful, but that is again a lot more effort. I will try to keep quiet.

gerhardol commented:
@RussKie 
I do not plan to make more changes to this (considering some enhancements, but that is for later)

RussKie commented on GitCommands/Git/GitRevisionTester.cs:
```diff
@@ -1,15 +1,106 @@
+            }
+
+            return true;
+        }
```
We need to rename the method to `AnyFirstAreParentsToSelected` to match the implementation and the doc "_True if **one of** the first selected is parent_".

RussKie replied:
:bulb:  Thinking of it, may be we can change the signature to something like the following? 

```
+        /// <summary>
+        /// Indicates whether any of the provided <param ref="selectedRevision" />'s parents are selected.
+        /// </summary>
+        public bool AnyRevisionParentsSelected(GitRevision selectedRevision, IEnumerable<GitRevision> revisionsSelectedFirst)
```

I think API becomes slightly more readable:

```
bool firstIsParent = _revisionDiffController.AnyRevisionParentsSelected(DiffFiles.Revision, DiffFiles.SelectedItemParents);
 ````



A rough idea, open to discussion...

gerhardol replied:
The method checks if "True if all of the first selected are parents" so this would be changed behavior

RussKie replied:
My bad, misread the logic.

gerhardol replied:
I will be happy to take a discussion why I have done like I have done and to change it to something better. How should this work? Your comment is a good indication of that. I could give some more background tomorrow. The alternative is to take this in follow-up discussions.

RussKie replied:
That's be good. I'm finding API somewhat confusing.

gerhardol replied:
This PR (including the related PRs) primarily improves the diff and presentation of versions. In this process, the menu logic has been updated, some alternatives were not relevant and other were missing. The logic was also changed so testing could be done (testing is for the display, not for the actions so I do not find the tests useful still but they suck less than they would in my first iterations).

For the alternatives, some kind of rules were used:
 * With one item selected, the menu items must be relevant
 * With multiple items selected, menu items are relevant if at least one item is relevant. For instance if submodules are available if at least one submodule is selected. (There are exceptions to this rule)

In general, the behavior should be the same if the parent-child is selected as if just the "child". It was not like this previously, selecting a parent-child was more limiting. 
In some situations selecting the parent, then the selected will provide more information, like the actual parents to parents, that enables those menu items (in diff menu).
AllFirstAreParentsToSelected() is a helper for these situations, this is similar to just selecting one revision.

There are some special handling for merge commits, to show the combined diff like in e51741dcbf694d937db2c0fbeabf107827562fa7. This selects all parents and adds the artificial DiffFiles.CombinedDiff to compare to the Selected revision. The combined diff is shown if only one of the parents to e517 is shown but would be confusing if a non parent was selected. 
(So combined diff is shown if at least one parent shown, but no other revision than parents).

For staged/unstaged, the availability is available is based on the selected revision but the parent must be staged/Head to be sure a file can be staged/unstaged. This commands breaks the rule to "show if at least one is possible" as it is not possible to determine the availability in a good way for the command.

FirstIsParent is also used to confirm that a parent to parent exists for difftool exists and to prune when parent-to-selected(B) is is the same as first(A).

Pushing a few clarifications related to this.

Not sure if it got any clearer...

RussKie commented on GitCommands/Git/GitRevisionTester.cs:
```diff
@@ -1,15 +1,106 @@
+
+            foreach (var item in items)
+            {
+                string filePath = _fullPathResolver?.Resolve(item.Name);
```
`_fullPathResolver` can't be `null`

gerhardol replied:
fixing

RussKie commented on GitUI/CommandsDialogs/FormDiff.cs:
```diff
@@ -65,6 +71,12 @@ public partial class FormDiff : GitModuleForm
 
+        protected override void OnRuntimeLoad(EventArgs e)
+        {
+            _revisionDiffContextMenuController = new FileStatusListContextMenuController();
```
:question:  Can this be moved into constructor? If so, this method an go....

gerhardol replied:
Yes, I just followed the pattern you had use in other modules.

RussKie replied:
I would have used `OnRuntimeLoad` instead of instantiating objects in constructor where would be a dependency on fully set `GitModule`, which isn't available until the control is loaded.
In this case there is no dependency on `GitModule` and I'm pretty sure it is safe to move the instantiation in to the constructor.

gerhardol replied:
Will move

RussKie commented on GitUI/CommandsDialogs/RevisionDiff.cs:
```diff
@@ -164,21 +155,56 @@ protected override void OnRuntimeLoad(EventArgs e)
+            return DescribeRevision(sha1, 0);
+        }
+
+        private string DescribeRevision(string sha1, int maxLength)
```
we can combine these two methods into one - `private string DescribeRevision(string sha1, int maxLength = 0)`

gerhardol replied:
changing

RussKie commented on GitUI/CommandsDialogs/RevisionDiff.cs:
```diff
@@ -164,21 +155,56 @@ protected override void OnRuntimeLoad(EventArgs e)
+        {
+            var parents = DiffFiles.SelectedItemParents
+                .Where(i => showUnstagedAndCombined ||
+                    !(i.Guid.IsNullOrWhiteSpace() || i.Guid == GitRevision.UnstagedGuid || i.Guid == DiffFiles.CombinedDiff.Text))
```
what does `i.Guid == DiffFiles.CombinedDiff.Text` mean?
`CombinedDiff.Text` is a localised string "Combined Diff", why are we comparing it to a commit's SHA1?

gerhardol replied:
DiffFiles.CombinedDiff.Text is used both for display and as an ID for a variant of an "artificial commit".

RussKie replied:
Something doesn't feel right about it.... can't articulate what exactly, but it feels bad

gerhardol replied:
It is not nice, but I preferred to not spread the knowledge about the revision. The proper way would be to add it in GitRevision.cs like:
        public const string IndexGuid = "1111111111111111111111111111111111111111";

RussKie replied:
we already have it
https://github.com/gitextensions/gitextensions/blob/663036a7fe10d928596580e0c849107b8d00ddf3/GitCommands/Git/GitRevision.cs#L18

gerhardol replied:
In a similar way:
public const string CombinedDiffGuid = "2222...";
DiffFiles.CombinedDiff.Text will be used in some situations still.

gerhardol replied:
Added a CombinedDiffGuid for the artificial diff after all. For instance #4154 seem to enhance the comparison more, so it is better to set the pattern. Not sure if the TranslationApp handles this correctly.

RussKie commented on UnitTests/GitCommandsTests/Git/GitRevisionTesterTests.cs:
```diff
@@ -10,12 +11,63 @@ namespace GitCommandsTests.Git
+        [Test]
+        public void LocalExists_should_return_false_if_none_of_locally_tracked_items_have_files()
+        {
+            Assert.Inconclusive();
```
you've done a great job with tests :+1: 

these tests need finishing too :wink:

gerhardol replied:
Will finish later, some done

gerhardol commented:
I have not planned to change anything more
